### PR TITLE
Remove ClusterImageSetMissing condition.

### PR DESCRIPTION
**Pull request by dgoodwin**
Move to RequirementsMet with reason ClusterImageSetMissing.

x-ref: https://issues.redhat.com/browse/HIVE-1481

### CODE REVIEW

**Comment by dgoodwin** [comment]
/assign @suhanime 
/cc @abhinavdahiya 

This is a first stab at implementing what we talked about in https://github.com/openshift/hive/pull/1413

File: pkg/controller/clusterdeployment/clusterdeployment_controller.go [hunk 1]
```diff
@@ -317,6 +316,35 @@ func (r *ReconcileClusterDeployment) Reconcile(ctx context.Context, request reco
...
+				legacy = true
+				break
+			}
+			if !legacy {
+				newConditions = append(newConditions, c)
+			} else {
+				cdLog.Infof("removing legacy condition: %v", c)
+			}
+		}
+	}
```
Comment (suhanime) [inline pkg/controller/clusterdeployment/clusterdeployment_controller.go:337]: Can't we use `FindClusterDeploymentCondition`? Looks like it is working [here](https://github.com/openshift/hive/pull/1430/files#diff-b9685aa0bf796d1db13613fe411ed287146777107405bcc935ebd1d45d64ea47R1457)

File: pkg/controller/clusterdeployment/clusterdeployment_controller_test.go [hunk 2]
```diff
@@ -2524,8 +2574,6 @@ func testEmptyClusterDeployment() *hivev1.ClusterDeployment {
 }
 
 func testInstallConfigSecret() *corev1.Secret {
-	encodedIC := base64.StdEncoding.EncodeToString([]byte(testAWSIC))
-	log.Info(encodedIC)
```
Comment (suhanime) [inline pkg/controller/clusterdeployment/clusterdeployment_controller_test.go:2528]: Doesn't look like it is related to these changes

**Review by suhanime — COMMENTED** [review]
Trajectory looks great.
Needs a `make vendor` and changing the tests which mention `hivev1.ClusterImageSetNotFoundCondition` in metrics and controllerutils tests

File: pkg/controller/clusterdeployment/clusterdeployment_controller.go (hunk 1 above)
Comment (dgoodwin) [reply pkg/controller/clusterdeployment/clusterdeployment_controller.go:337]: That would let me find if it's there, but we have no delete function and probably shouldn't given we want them always there (except for legacy). Reslicing looked a little clunk, and requires an index, which FindClusterDeploymentCondition doesn't expose.  Given all this it felt more reasonable to just iterate the slice and built up a new with the ones we want to keep. Let me know if I missed something obvious though.

File: pkg/controller/clusterdeployment/clusterdeployment_controller_test.go (hunk 2 above)
Comment (dgoodwin) [reply pkg/controller/clusterdeployment/clusterdeployment_controller_test.go:2528]: It's not it's from a recent PR I did, but doing a separate PR and CI cycle to remove two lines of logging feels worse to me than including it in a diff PR. WDYT?

File: pkg/controller/clusterdeployment/clusterdeployment_controller.go [hunk 3]
```diff
@@ -1086,26 +1115,38 @@ func (r *ReconcileClusterDeployment) setDeprovisionLaunchErrorCondition(cd *hive
 }
 
 func (r *ReconcileClusterDeployment) setImageSetNotFoundCondition(cd *hivev1.ClusterDeployment, name string, isNotFound bool, cdLog log.FieldLogger) error {
```
[the remark this answers is not in the document]
Comment (dgoodwin) [reply pkg/controller/clusterdeployment/clusterdeployment_controller.go:1117]: renamed to setReqsMetConditionImageSetNotFound

File: pkg/controller/clusterdeployment/clusterdeployment_controller.go [hunk 4]
```diff
@@ -1086,26 +1115,38 @@ func (r *ReconcileClusterDeployment) setDeprovisionLaunchErrorCondition(cd *hive
...
-		message,
-		controllerutils.UpdateConditionNever)
 	if !changed {
 		return nil
 	}
-	cdLog.Infof("setting ClusterImageSetNotFoundCondition to %v", status)
 	cd.Status.Conditions = conds
+	reqsCond := controllerutils.FindClusterDeploymentCondition(cd.Status.Conditions,
+		hivev1.RequirementsMetCondition)
+	cdLog.Info("updating RequirementsMetCondition: %v", reqsCond)
```
[the remark this answers is not in the document]
Comment (dgoodwin) [reply pkg/controller/clusterdeployment/clusterdeployment_controller.go:1149]: Updated

File: pkg/controller/clusterdeployment/clusterdeployment_controller.go (hunk 1 above)
Comment (suhanime) [reply pkg/controller/clusterdeployment/clusterdeployment_controller.go:337]: The logic is good - I just meant to see if we could make it not clunky - it is fine this way...

File: pkg/controller/clusterdeployment/clusterdeployment_controller_test.go (hunk 2 above)
Comment (suhanime) [reply pkg/controller/clusterdeployment/clusterdeployment_controller_test.go:2528]: No worries - just wanted to make sure it was purposeful

**Review by suhanime — COMMENTED** [review]
metrics and controllerutils tests need to be fixed post vendor changes, everything else lgtm
` vet: pkg/controller/metrics/metrics_test.go:35:12: ClusterImageSetNotFoundCondition not declared by package v1
# github.com/openshift/hive/pkg/controller/utils
vet: pkg/controller/utils/conditions_test.go:15:10: ClusterImageSetNotFoundCondition not declared by package v1
make: *** [verify-govet] Error 2 `

**Comment by dgoodwin** [comment]
Updated, thanks for looking!

File: pkg/controller/clusterdeployment/clusterdeployment_controller.go [hunk 5]
```diff
@@ -317,6 +316,35 @@ func (r *ReconcileClusterDeployment) Reconcile(ctx context.Context, request reco
...
+				break
+			}
+			if !legacy {
+				newConditions = append(newConditions, c)
+			} else {
+				cdLog.Infof("removing legacy condition: %v", c)
+			}
+		}
+	}
+	if len(newConditions) < len(cd.Status.Conditions) {
```
Comment (2uasimojo) [inline pkg/controller/clusterdeployment/clusterdeployment_controller.go:338]: Is this future-proof? Would it be better to use a bool flag set to `true` in the loop above? (Like `legacy`, but wider scoped.)

File: pkg/controller/clusterdeployment/clusterdeployment_controller.go [hunk 6]
```diff
@@ -317,6 +316,35 @@ func (r *ReconcileClusterDeployment) Reconcile(ctx context.Context, request reco
...
+			if string(c.Type) == lc {
+				legacy = true
+				break
+			}
+			if !legacy {
+				newConditions = append(newConditions, c)
+			} else {
+				cdLog.Infof("removing legacy condition: %v", c)
+			}
+		}
```
Comment (2uasimojo) [inline pkg/controller/clusterdeployment/clusterdeployment_controller.go:336]:
```suggestion
		legacy := false
		for _, lc := range legacyConditions {
			if string(c.Type) == lc {
				legacy = true
				break
			}
		}
		if !legacy {
			newConditions = append(newConditions, c)
		} else {
			cdLog.Infof("removing legacy condition: %v", c)
		}
```
?

File: pkg/controller/clusterdeployment/clusterdeployment_controller.go [hunk 7]
```diff
@@ -317,6 +316,35 @@ func (r *ReconcileClusterDeployment) Reconcile(ctx context.Context, request reco
...
+	}
+	if len(newConditions) < len(cd.Status.Conditions) {
+		cd.Status.Conditions = newConditions
+		cdLog.Info("removed legacy conditions")
```
Comment (2uasimojo) [inline pkg/controller/clusterdeployment/clusterdeployment_controller.go:340]: Seems unnecessary with the log(s) on L334.

File: pkg/controller/clusterdeployment/clusterdeployment_controller.go (hunk 6 above)
Comment (dgoodwin) [reply pkg/controller/clusterdeployment/clusterdeployment_controller.go:336]: Whoops!

File: pkg/controller/clusterdeployment/clusterdeployment_controller.go (hunk 5 above)
Comment (dgoodwin) [reply pkg/controller/clusterdeployment/clusterdeployment_controller.go:338]: Hmm perhaps if someone injected some code in the middle of these statements. I'll add a bool.

**Comment by dgoodwin** [comment]
Updates pushed.

**Pull request merged**